mempool/src/tx_origin.rs

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use p2p_types::peer_id::PeerId;
+
+/// Tracks where a transaction originates
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy)]
+pub enum TxOrigin {
+    /// Transaction was submitted to local node's mempool. It should not be propagated further.
+    LocalMempool,
+
+    /// Transaction was submitted via local node's RPC subsystem. It should be propagated if valid.
+    LocalP2p,
+
+    /// Transaction was in a block but moved into the mempool upon a reorg.
+    PastBlock,
+
+    /// Transaction was received from a peer.
+    ///
+    /// If it eventually turns out to be valid, it should be propagated further to other peers.
+    /// If it's not valid, the original peer should be penalized as appropriate.
```

Comment:
I just want to say that there are interesting implications to wanting to use the peer id for extended periods of time. I think currently in the p2p system, if a peer connects and then disconnects and then connects again, they will get a new peer id. The peer id is not associated with an entity, it's just an atomic counter for connections. So this can be easily circumvented. 

I believe the peer id is the same in bitcoin core. It doesn't have any identity associated with it.

There are multiple solutions to this problem, including creating a special type in p2p that's associated with a peer, truly.

Reply:
> I believe the peer id is the same in bitcoin core. It doesn't have any identity associated with it.

I just checked and indeed it is.